Add supertokens intergration (#1)
jscyo/supabase: web/docs/guides/integrations/supertokens.mdx

In this guide we will build a simple web application using SuperTokens, Supabase, and Next.js. You will be able to sign up and your email will be stored in Supabase. Once authenticated the frontend will be able to query Supabase and retrieve the user's email. Our example app will be using the [Email-Password and Social Login](https://supertokens.com/docs/thirdpartyemailpassword/introduction) recipe for authentication and session management.

We will use Supabase to store and authorize access to user data. Supabase makes it simple to setup Row Level Security(RLS) policies  which ensure users can only read and write data they are authorized to do so. If you do not have a Supabase account, you will need to create one.

> Note: You can find a completed version of this project on [Github](https://github.com/supertokens/supertokens-auth-react/tree/master/examples/with-supabase)

## Step 1: Create a new Supabase project
From your [Supabase dashboard](https://app.supabase.io/), click `New project`.

Enter a `Name` for your Supabase project.

Enter a secure `Database Password`.

> Select the same `Region` you selected for your Auth0 tenant.

Reviewer: ```suggestion
Select the same `Region` you host your app's backend in.
```

Click `Create new project`.

![New Supabase project settings](/img/guides/integrations/supertokens/supabase_dashboard_create.png)

## Step 2: Creating tables in Supabase
From the sidebar menu in the [Supabase dashboard](https://app.supabase.io/), click `Table editor`, then `New table`.

Enter `users` as the `Name` field.

Select `Enable Row Level Security (RLS)`.

Remove the default columns

Create two new columns:

- `user_id` as `varchar` as primary key
- `email` as `varchar`

Click `Save` to create the new table.

![Users table](/img/guides/integrations/supertokens/supabase_table_create.png)

## Step 3: Create Policies to allow Select and Insert

> Our polices will need to know who our currently logged in user is to determine whether or not they should have access. Let's create a PostgreSQL function to extract the current user from our new JWT.

Reviewer: You can give an example of the KWT we created (the contents of it)

Navigate back to the Supabase dashboard, select `SQL` from the sidebar menu, and click `New query`. This will create a new query called `new sql snippet`, which will allow us to run any SQL against our Postgres database.

Write the following and click `Run`.

```sql
create or replace function auth.user_id() returns text as $$
  select nullif(current_setting('request.jwt.claims', true)::json->>'userId', '')::text;
$$ language sql stable;
```

This will create a function called `auth.user_id()`, which will inspect the `userId` field of our JWT payload.

> Let's create our first policy that checks whether this user is the owner of the email.

Reviewer: Put this under a heading like `SELECT query policy`

                            createNewSession: async function (input) {
                                const payload = {
                                    userId: input.userId,
                                    exp: Math.floor(Date.now() / 1000) + 60 * 60,
                                };

                                const supabase_jwt_token = jwt.sign(payload, process.env.SUPABASE_SIGNING_SECRET);

                                input.accessTokenPayload = {
                                    ...input.accessTokenPayload,
                                    supabase_token: supabase_jwt_token,
                                };

                                return await originalImplementation.createNewSession(input);
                            },
                        };
                    },
                },
            }),
        ],
        isInServerlessEnv: true,
    };
};

```

As seen above, we will be using the `jsonwebtoken` library to create a JWT signed with Supabase's signing secret whose payload contains the user's userId.

> We will be storing this token in the `accessTokenPayload` which will essentially allow us to access the JWT whenever the user's session is accessible.

Reviewer: ```suggestion
We will be storing this token in the `accessTokenPayload` which will essentially allow us to access the `supabase_token` on the frontend and backend whilst the user is logged in.
```

## Step 6: Requesting data from Supabase 

Create a new file called `utils/supabase.ts` and add the following:

```ts
// utils/supabase.ts

import { createClient } from "@supabase/supabase-js";

const getSupabase = (access_token) => {
    const supabase = createClient(process.env.NEXT_PUBLIC_SUPABASE_URL, process.env.NEXT_PUBLIC_SUPABASE_KEY);

    supabase.auth.session = () => ({
        access_token
    });

    return supabase;
};

export { getSupabase };
```

> This will be our client for talking to Supabase. We can pass it an `access_token` and it will be attached to our request.

Reviewer: ```suggestion
This will be our client for talking to Supabase. We can pass it an `access_token` and it will be attached to our request. This `access_token` is the same as the `supabase_token` we had created earlier.
```

## Step 7: Overriding SuperTokens Sign Up APIs:

In our example app there are two ways for signing up a user. Email-Password and Social Login based authentication. We will need to override both these APIs such that when a user signs up, their email mapped to their userId is stored in Supabase.

```ts
// config/backendConfig.ts

import ThirdPartyEmailPasswordNode from "supertokens-node/recipe/thirdpartyemailpassword";
import SessionNode from "supertokens-node/recipe/session";
import { TypeInput } from "supertokens-node/lib/build/types";
import { appInfo } from "./appInfo";
import jwt from "jsonwebtoken";
import { getSupabase } from "../utils/supabase";

let backendConfig = (): TypeInput => {
    return {
        framework: "express",
        supertokens: {
            connectionURI: "https://try.supertokens.com",
        },
        appInfo,
        recipeList: [
            ThirdPartyEmailPasswordNode.init({

>                 providers: [
>                     // We have provided you with development keys which you can use for testing.
>                     // IMPORTANT: Please replace them with your own OAuth keys for production use.
>                     ThirdPartyEmailPasswordNode.Google({
>                         clientId: process.env.GOOGLE_CLIENT_ID,
>                         clientSecret: process.env.GOOGLE_CLIENT_SECRET,
>                     }),
>                     ThirdPartyEmailPasswordNode.Github({
>                         clientId: process.env.GITHUB_CLIENT_ID,
>                         clientSecret: process.env.GITHUB_CLIENT_SECRET,
>                     }),
>                     ThirdPartyEmailPasswordNode.Apple({
>                         clientId: process.env.APPLE_CLIENT_ID,
>                         clientSecret: {
>                             keyId: process.env.APPLE_KEY_ID,
>                             privateKey: process.env.APPLE_PRIVATE_KEY.replace(/\\n/g, "\n"),
>                             teamId: process.env.APPLE_TEAM_ID,
>                         },
>                     }),
>                 ],

Reviewer: ```suggestion
                providers: [...],
```

                override: {
                    apis: (originalImplementation) => {
                        return {
                            ...originalImplementation,

>                             thirdPartySignInUpPOST: async function (input) {

Reviewer: Add comments explaining what these functions are

                                }
                                let response = await originalImplementation.thirdPartySignInUpPOST(input);

                                if (response.status === "OK" && response.createdNewUser) {
                                    // retrieve the supabase_token from the accessTokenPayload
                                    const accessTokenPayload = response.session.getAccessTokenPayload();

                                    // create a supabase client whose JWT contains the user's id
                                    const supabase = getSupabase(accessTokenPayload.supabase_token);
                                    // store the user's email mapped to their userId in Supabase
                                    const { error } = await supabase
                                        .from("users")
                                        .insert({ email: response.user.email, user_id: response.user.id });

                                    if (error !== null) {
                                        // Since Row Level Security is enabled in our Supabase tables, if a policy for inserting

>                                         // rows to a table has not been defined, insertion will throw an error.

Reviewer: Remove this comment

                                    throw Error("Should never come here");
                                }

                                let response = await originalImplementation.emailPasswordSignUpPOST(input);

                                if (response.status === "OK") {
                                    // retrieve the supabase_token from the accessTokenPayload
                                    const accessTokenPayload = response.session.getAccessTokenPayload();

                                    // create a supabase client whose JWT contains the user's id
                                    const supabase = getSupabase(accessTokenPayload.supabase_token);

                                    // store the user's email mapped to their userId in Supabase
                                    const { error } = await supabase
                                        .from("users")
                                        .insert({ email: response.user.email, user_id: response.user.id });

                                    if (error !== null) {
                                        // Since Row Level Security is enabled in our Supabase tables, if a policy for inserting

>                                         // rows to a table has not been defined, insertion will throw an error.

Reviewer: Remove comment

                            // it can be used by Supabase to validate the user when retrieving user data from their service.
                            // We store this token in the accessTokenPayload so it can be accessed on the frontend and on the backend.
                            createNewSession: async function (input) {
                                const payload = {
                                    userId: input.userId,
                                    exp: Math.floor(Date.now() / 1000) + 60 * 60,
                                };

                                const supabase_jwt_token = jwt.sign(payload, process.env.SUPABASE_SIGNING_SECRET);

                                input.accessTokenPayload = {
                                    ...input.accessTokenPayload,
                                    supabase_token: supabase_jwt_token,
                                };

                                return await originalImplementation.createNewSession(input);
                            },
                        };
                    },
                },
            }),
        ],
        isInServerlessEnv: true,
    };
};

```

> As seen above, we will be overriding the `emailPasswordSignUpPOST` and `thirdPartySignInUpPOST` APIs such that if a user signs up, we retrieve the Supabase JWT (which we created in the `createNewSession` function and attached to the user's session) from the user's accessTokenPayload and send a request to Supabase to insert the email-userid mapping.

Reviewer: ```suggestion
As seen above, we will be overriding the `emailPasswordSignUpPOST` and `thirdPartySignInUpPOST` APIs such that if a user signs up, we retrieve the Supabase JWT (which we created in the `createNewSession` function) from the user's accessTokenPayload and send a request to Supabase to insert the email-userid mapping.
```
